security-command-center/snippets/src/main/java/vtwo/client/CreateClientWithEndpoint.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vtwo.client;
+
+// [START securitycenter_set_client_endpoint_v2]
+
+import com.google.cloud.securitycenter.v2.SecurityCenterClient;
+import com.google.cloud.securitycenter.v2.SecurityCenterSettings;
+import java.io.IOException;
+
+public class CreateClientWithEndpoint {
+
+  public static void main(String[] args) throws IOException {
+    // TODO: Replace the value with your regional endpoint.
+    String regionalEndpoint = "securitycenter.REGION.rep.googleapis.com:443";
+
+    createClientWithEndpoint(regionalEndpoint);
+  }
+
+  // Creates Security Command Center clients for the default endpoint and for a
+  // regional endpoint.
+  public static void createClientWithEndpoint(String regionalEndpoint) throws java.io.IOException {
+    SecurityCenterSettings regionalSettings =
+        SecurityCenterSettings.newBuilder().setEndpoint(regionalEndpoint).build();
+
+    try (SecurityCenterClient client = SecurityCenterClient.create();
+        SecurityCenterClient regionalClient = SecurityCenterClient.create(regionalSettings)) {
```

Comment:
Why a user has to create two instances of `SecurityCenterClient`?

The code sample method is called "createClientWithEndpoint". Please, remove the client creation without regional settings.

Reply:
There are a couple of reasons why the sample needs to create both clients:

* Security Command Center's regional endpoint supports a limited subset of methods in the API. If you're using a regional endpoint, you'll also have to use the global endpoint.

    This behavior is pretty unusual—I'm not aware of any other Google Cloud service that requires customers to do this. Given how unusual this is, I'd like the code sample to make it very clear that customers must create two clients.

* I'd like the Java code sample to match the code samples for other languages.

Does that work for you?

Reply:
I am unsure that I understand. Code example _is defined_ as "creating client with **regional** endpoint".

If the following code does not create **regional** endpoint why this code is included in the example?

```java
SecurityCenterClient client = SecurityCenterClient.create();
```

Your explanation does not sit well with me. Do you say that in order to use **regional** endpoint of SCC a developer has to create another (what it is called, global?) instance of `SecurityCenterClient`? What API endpoint does this "default" (or global) instance calls?

If this code sample confuses me, there is a good chance it will confuse an ordinary developer who has no experience about API endpoints and other specifics.

To your question:

> I'm not aware of any other Google Cloud service

AFAIK majority of services implement regional endpoints. Easy examples are GCS and PubSub. There are plenty more.
Additionally, many client libraries require endpoint (usually in the form of URL) parameter to initialize the object.

Regarding the following:

> I'd like the Java code sample to match the code samples for other languages.

This is a good goal when the code sample is clear and not ambiguous. Unfortunately, this is not the case with this particular code sample.

Reply:
> Your explanation does not sit well with me. Do you say that in order to use **regional** endpoint of SCC a developer has to create another (what it is called, global?) instance of `SecurityCenterClient`? What API endpoint does this "default" (or global) instance calls?
>
> ...
>
> AFAIK majority of services implement regional endpoints. Easy examples are GCS and PubSub. There are plenty more.

Let me see if I can explain this more clearly:

Yes, a number of Google Cloud services provide regional endpoints. Let's use [Cloud DLP](https://cloud.google.com/sensitive-data-protection/docs/reference/rest) as an example, because it's a pretty conventional service. (In contrast, Cloud Storage is an outlier in many different ways, and the only publicly visible endpoint for Cloud Pub/Sub [appears to be the global endpoint](https://cloud.google.com/pubsub/docs/reference/rest).)

As far as I can tell, Cloud DLP lets you use _any_ endpoint to call _any_ method in the API. In other words, if you use a regional service endpoint, and all of your data is in that region, you'll never have to call any other endpoint. If there are exceptions, then the [relevant documentation](https://cloud.google.com/sensitive-data-protection/docs/api-endpoints) doesn't mention them.

Security Command Center doesn't work that way:

* A [subset of Security Command Center resource types](https://cloud.google.com/security-command-center/docs/data-residency-support#scc-resources-subject-to-drz) are subject to data residency controls. If your data resides in a location with a regional endpoint (currently just `me-central2`), then you _must_ use the regional endpoint to manage these resource types.
* Other resource types _are not_ subject to data residency controls. As a result, you _cannot_ use a regional endpoint to manage other resource types. You must use the global endpoint.

As a result, if you enable data residency for Security Command Center, and your data resides in `me-central2`, then you _must_ create two separate clients:

* One for the regional endpoint, `https://securitycenter.me-central2.rep.googleapis.com`, to manage resources that reside in `me-central2`
* One for the global endpoint, `https://securitycenter.googleapis.com`, to manage resources that aren't subject to data residency controls

I'm not aware of any other Google Cloud services that require you to do this. That's why we need to provide a code sample—as you've pointed out, this requirement is surprising, and this isn't a pattern that our customers are already familiar with.

Does that make more sense? If not, what would you like me to change? (Keep in mind that customers are likely to see this sample in the context of the [data residency documentation](https://cloud.google.com/security-command-center/docs/data-residency-support), which already provides some of this context.)

> Additionally, many client libraries require endpoint (usually in the form of URL) parameter to initialize the object.

Out of curiosity, which ones? I don't think I've seen any that _require_ an endpoint parameter.

Reply:
resolving this comment following our conversation. please, see two suggestions